File: python/Q105_construct_binary_tree_from_preorder_and_inorder_traversal.py
#
# @lc app=leetcode id=105 lang=python3
#
# [105] Construct Binary Tree from Preorder and Inorder Traversal
#


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None) -> None:
        self.val: int = val
        self.left: TreeNode | None = left
        self.right: TreeNode | None = right


# @lc code=start
class Solution:
    def build_tree(self, limit: int) -> TreeNode | None:
        if self.pre_index >= len(self.preorder):
            return None

        if self.inorder[self.in_index] == limit:
            self.in_index += 1
            return None

        value: int = self.preorder[self.pre_index]

        new_node: TreeNode = TreeNode(value)
        self.pre_index += 1

        new_node.left = self.build_tree(value)
        new_node.right = self.build_tree(limit)

        return new_node

    def buildTree(self, preorder: list[int], inorder: list[int]) -> TreeNode | None:
        self.preorder: list[int] = preorder
        self.inorder: list[int] = inorder

        self.pre_index: int = 0
        self.in_index: int = 0

        return self.build_tree(3001)


# @lc code=end

File: rust/src/leetcode/Q105_construct_binary_tree_from_preorder_and_inorder_traversal.rs
/*
 * @lc app=leetcode id=105 lang=rust
 *
 * [105] Construct Binary Tree from Preorder and Inorder Traversal
 */

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// @lc code=start
use std::cell::RefCell;
use std::collections::HashMap;
use std::rc::Rc;

impl Solution {
    fn build_node(
        preorder: &Vec<i32>,
        inorder_map: &HashMap<i32, i32>,
        preorder_index: &mut usize,
        left: i32,
        right: i32,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        if left > right {
            return None;
        }

        let value: i32 = preorder[*preorder_index];
        let node: Rc<RefCell<TreeNode>> = Rc::new(RefCell::new(TreeNode::new(value)));
        *preorder_index += 1;

        if left == right {
            return Some(node);
        }

        let inorder_index: i32 = inorder_map[&value];

        node.borrow_mut().left = Self::build_node(
            preorder,
            inorder_map,
            preorder_index,
            left,
            inorder_index - 1,
        );
        node.borrow_mut().right = Self::build_node(
            preorder,
            inorder_map,
            preorder_index,
            inorder_index + 1,
            right,
        );

        Some(node)
    }

    pub fn build_tree(preorder: Vec<i32>, inorder: Vec<i32>) -> Option<Rc<RefCell<TreeNode>>> {
        let mut inorder_map: HashMap<i32, i32> = HashMap::new();
        for (index, value) in inorder.iter().enumerate() {
            inorder_map.entry(*value).or_insert(index as i32);
        }

        let mut preorder_index: usize = 0;

        Self::build_node(
            &preorder,
            &inorder_map,
            &mut preorder_index,
            0,
            preorder.len() as i32 - 1,
        )
    }
}
// @lc code=end
